croaring/src/bitset/imp.rs

```diff
@@ -0,0 +1,479 @@
+use super::Bitset;
+use std::{mem, ptr};
+
+impl Bitset {
+    #[inline]
+    #[allow(clippy::assertions_on_constants)]
+    pub(super) unsafe fn take_heap(p: *mut ffi::bitset_t) -> Self {
+        assert!(!p.is_null());
+        let result = Self { bitset: p.read() };
+        // It seems unlikely that the bitset type will meaningfully change, but check if we ever go
+        // to a version 2.
+        const _: () = assert!(ffi::ROARING_VERSION_MAJOR == 1);
+        ffi::roaring_free(p.cast());
+        result
+    }
+
+    /// Access the raw underlying slice
+    #[inline]
+    pub const fn as_slice(&self) -> &[u64] {
+        if self.bitset.arraysize == 0 {
+            &[]
+        } else {
+            unsafe { std::slice::from_raw_parts(self.bitset.array, self.bitset.arraysize) }
+        }
+    }
+
+    /// Access the raw underlying slice
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [u64] {
+        if self.bitset.arraysize == 0 {
+            &mut []
+        } else {
+            unsafe { std::slice::from_raw_parts_mut(self.bitset.array, self.bitset.arraysize) }
+        }
+    }
+
+    /// Create a new bitset
+    ///
+    /// Does not allocate
+    ///
+    /// # Examples
+    /// ```
+    /// use croaring::Bitset;
+    /// let bitset = Bitset::new();
+    /// assert_eq!(bitset.capacity(), 0);
+    /// assert_eq!(bitset.size_in_bits(), 0);
+    /// ```
+    #[inline]
+    #[doc(alias = "bitset_create")]
+    pub const fn new() -> Self {
+        Self {
+            bitset: ffi::bitset_t {
+                array: ptr::null_mut(),
+                arraysize: 0,
+                capacity: 0,
+            },
+        }
+    }
+
+    /// Create a new bitset of the specified size
+    ///
+    /// # Examples
+    /// ```
+    /// use croaring::Bitset;
+    /// let bitset = Bitset::with_size(100);
+    /// // Actual size/capacity may be rounded up
+    /// assert!(bitset.capacity() >= 100);
+    /// assert!(bitset.size_in_bits() >= 100);
+    /// ```
+    #[inline]
+    #[doc(alias = "bitset_create_with_capacity")]
+    pub fn with_size(size: usize) -> Self {
+        unsafe { Self::take_heap(ffi::bitset_create_with_capacity(size)) }
+    }
+
+    /// Capacity in bits
+    #[inline]
+    pub const fn capacity(&self) -> usize {
+        self.bitset.capacity * 64
+    }
+
+    /// Set all bits to zero
+    ///
+    /// # Examples
+    /// ```
+    /// use croaring::Bitset;
+    /// let mut bitset = Bitset::with_size(64);
+    /// bitset.fill(true);
+    /// assert!(bitset.get(1));
+    /// assert!(bitset.get(63));
+    /// // Bitset size stays the same
+    /// assert!(!bitset.get(64));
+    /// bitset.fill(false);
+    /// assert!(!bitset.get(1));
+    /// assert!(!bitset.get(63));
+    /// assert!(!bitset.get(64));
+    /// ```
+    #[inline]
+    #[doc(alias = "bitset_clear")]
+    #[doc(alias = "bitset_fill")]
+    pub fn fill(&mut self, value: bool) {
```

Comment:
Not sure I like this. Any reason why `bitset_clear` and `bitset_fill` aren't separate?

Reply:
My thoughts were to be closer to [`<[bool]>::fill`](https://doc.rust-lang.org/std/primitive.slice.html#method.fill), and that since it's inline, `.fill(constant)` will be optimized to a single call without a branch anyway.

I really don't want to use the name `clear` (since it's too similar to [`Vec::clear`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.clear) but does something VERY different), so the alternative in my mind would be `Bitset::fill_zeros` and `Bitset::fill_ones`.

I'm not hard set on this option, but it seemed like the best option to me.